### PR TITLE
Adds support for tables, infoboxes and youtube in MD files

### PR DESCRIPTION
**Pull request by petele**
I'd like to use this for WebFundamentals when we export code labs and other files

### CODE REVIEW

File: claat/render/md.go [hunk 1]
```diff
@@ -232,3 +233,37 @@ func (mw *mdWriter) header(n *types.HeaderNode) {
...
+		}
+		mw.writeString("\n</tr>")
+	}
+	mw.writeString("</table>")
+}
+
+func (mw *mdWriter) infobox(n *types.InfoboxNode) {
+	mw.newBlock()
+	mw.writeString(`<aside markdown="1" class="`)
+	mw.writeString(string(n.Kind))
```
Comment (x1ddos) [inline claat/render/md.go]: there's also mw.writeBytes

File: claat/render/md.go [hunk 2]
```diff
@@ -232,3 +233,37 @@ func (mw *mdWriter) header(n *types.HeaderNode) {
...
+	mw.writeString("</aside>")
+}
+
+func (mw *mdWriter) youtube(n *types.YouTubeNode) {
+	mw.newBlock()
+	mw.writeString(`<div class="video-wrapper">\n`)
+	mw.writeString(`  <iframe class="devsite-embedded-youtube-video" `)
+	mw.writeString(`data-autohide="1" data-showinfo="0" frameborder="0" `)
+	mw.writeString(`allowfullscreen data-video-id="`)
+	mw.writeString(string(n.VideoID))
```
Comment (x1ddos) [inline claat/render/md.go]: mw.writeBytes

File: claat/render/md.go [hunk 3]
```diff
@@ -232,3 +233,37 @@ func (mw *mdWriter) header(n *types.HeaderNode) {
...
+	mw.writeString(string(n.Kind))
+	mw.writeString(`">`)
+	mw.write(n.Content.Nodes...)
+	mw.writeString("</aside>")
+}
+
+func (mw *mdWriter) youtube(n *types.YouTubeNode) {
+	mw.newBlock()
+	mw.writeString(`<div class="video-wrapper">\n`)
+	mw.writeString(`  <iframe class="devsite-embedded-youtube-video" `)
```
Comment (x1ddos) [inline claat/render/md.go]: not convinced by the "deviste-" prefix... would it be possible to drop? i.e. just `embedded-youtube-video`.

File: claat/render/md.go (hunk 2 above)
Comment (x1ddos) [inline claat/render/md.go]: better yet, merged with the line above:

```go
mw.writeFmt("data-video-id=%q", n.VideoID)
```

in which case I would also move "allowfullscreen" part to the line above it.

File: claat/render/md.go (hunk 1 above)
Comment (x1ddos) [inline claat/render/md.go]: Actually, you could just use `mw.writeFmt("<... class=%q>", n.Kind)`

But, because it's HTML now, you'll have to make sure it's safe using escaping. See this, for instance:
https://github.com/googlecodelabs/tools/blob/508f3fab2b82b08747277d730570444f1ab2c444/claat/render/html.go#L135

File: claat/render/md.go [hunk 4]
```diff
@@ -159,11 +164,7 @@ func (mw *mdWriter) url(n *types.URLNode) {
 	if n.URL != "" {
 		mw.writeString("[")
 	}
-	for _, cn := range n.Content.Nodes {
-		if t, ok := cn.(*types.TextNode); ok {
-			mw.writeString(t.Value)
-		}
-	}
+	mw.write(n.Content.Nodes...)
```
Comment (x1ddos) [inline claat/render/md.go]: Is this to make the "download" button a non-empty link text?

Comment (petele) [inline claat/render/md.go]: Yep, otherwise it is just an empty block right now

File: claat/parser/gdoc/parse.go [hunk 5]
```diff
@@ -729,10 +729,14 @@ func link(ds *docState) types.Node {
 	if ds.flags&fMakeCode != 0 || isCode(ds.css, ds.cur.Parent) {
 		t.Code = true
 	}
-	if href == "" || href[0] == '#' {
+	if href == "" {
 		t.MutateBlock(findBlockParent(ds.cur))
 		return t
 	}
+  if href[0] == '#' && href[1] == 'h' && href[2] == '.' {
```
Comment (x1ddos) [inline claat/parser/gdoc/parse.go]: why this change?

File: claat/render/md.go [hunk 6]
```diff
@@ -209,7 +215,7 @@ func (mw *mdWriter) list(n *types.ListNode) {
 }
 
 func (mw *mdWriter) itemsList(n *types.ItemsListNode) {
-	mw.newBlock()
+  mw.newBlock()
```
Comment (x1ddos) [inline claat/render/md.go]: this seems unnecessary.

File: claat/render/md.go [hunk 7]
```diff
@@ -232,3 +238,33 @@ func (mw *mdWriter) header(n *types.HeaderNode) {
...
+  mw.writeString(`<aside markdown="1" class="`)
+  mw.writeEscape(string(n.Kind))
+  mw.writeString(`">`)
+  mw.write(n.Content.Nodes...)
+  mw.writeString("</aside>")
+}
+
+func (mw *mdWriter) youtube(n *types.YouTubeNode) {
+  mw.newBlock()
+  mw.writeFmt("<iframe src=\"https://www.youtube.com/embed/%q\" ", n.VideoID)
```
Comment (x1ddos) [inline claat/render/md.go]: 1. you can use backquotes. it reads better:

    mw.writeFmt(`<iframe src="..." `, n.VideoID)

2. the "%q" wraps string in quotes. I don't think it's what you want. you probably meant "%s"
3. are we sure n.VideoID is safe? if in doubt, I'd use [url.PathEscape](https://golang.org/pkg/net/url/#PathEscape)

File: claat/render/md.go [hunk 8]
```diff
@@ -16,6 +16,8 @@ package render
 
 import (
 	"bytes"
+	"fmt"
+  htmlTemplate "html/template"
```
Comment (x1ddos) [inline claat/render/md.go]: I think you can import it as is, without renaming: import ( "html/template" )
there's no other "template" in this file, is there?

File: claat/render/md.go [hunk 9]
```diff
@@ -232,3 +238,33 @@ func (mw *mdWriter) header(n *types.HeaderNode) {
...
+    mw.writeString("\n</tr>")
+  }
+  mw.writeString("</table>")
+}
+
+func (mw *mdWriter) infobox(n *types.InfoboxNode) {
+  mw.newBlock()
+  mw.writeString(`<aside markdown="1" class="`)
+  mw.writeEscape(string(n.Kind))
+  mw.writeString(`">`)
```
Comment (x1ddos) [inline claat/render/md.go]: maybe the above three lines could be merged into one. seems simpler:

```go
mw.writeFmt(`<aside markdown="1" class="%s">`, template.HTMLEscapeString(n.Kind))
```

**Comment by marcacohen** [comment]
Hi @petele . Any chance to follow through on this? If you're busy, I can run with it. I'd like to see this get added, especially the video embedding.

**Comment by josephkandi** [comment]
What happened to this? Any chance this will go through?

File: claat/render/md.go (hunk 4 above)
Comment (baoduy1103) [reply claat/render/md.go:172]: relly ??
